source/type/pt_data.h

```diff
@@ -59,8 +59,16 @@ namespace type {
 typedef std::map<std::string, std::string> code_value_map_type;
 typedef std::map<std::string, code_value_map_type> type_code_codes_map_type;
 struct PT_Data : boost::noncopyable{
-#define COLLECTION_AND_MAP(type_name, collection_name) std::vector<type_name*> collection_name; std::unordered_map<std::string, type_name *> collection_name##_map;
+    template<typename T> const std::vector<T*>& collection() const {
+        static_assert(!std::is_same<T, T>::value, "PT_Data::collection() not implemented");
+        static const std::vector<T*> collection;
+        return collection;
```

Comment:
i dont know if return {} will compile? you don't have to change it in this pr, i am just curious about it